days/day-09/solutions/day09.stektpotet.py

```diff
@@ -0,0 +1,25 @@
+import fileinput
+import itertools
+
+if __name__ == '__main__':
+    lines = [int(line) for line in fileinput.input()]
+    magic_number = None
+    for p in range(len(lines) - 26):
+        if not any(s == lines[p+25] for s in map(sum, itertools.combinations(lines[p: p+25], 2))):
+            magic_number = lines[p + 25]
+            print(magic_number)
+            break
+
+    for i, a in enumerate(lines):
+        s = a
+        r_min = a
+        r_max = a
+        for j, b in zip(range(i+1, len(lines)), lines[i+1:]):
+            s += b
+            r_min = min(r_min, b)
+            r_max = max(r_max, b)
+            if s > magic_number:
+                break
```

Comment:
Aha najs optimization :clap:

Reply:
```py
if not any(s == lines[p+25] for s in map(sum, itertools.combinations(lines[p: p+25], 2))):
```
This one is pretty cool too imo:

because `any` accepts iterators, hence also generators or generator expressions, such as above, it will not exhaust the for loop when `s==lines[p+25]` is met for any single instance of s. `map`also behaves the same way, meaning that I'm effectively...
1. not summing any more combinations than neccessary, and
2. not creating any more combinations from the preamable.

making that line more effective than my original loop:

```py	    
for p in range(len(lines) - 26):
    found=False
    for a, b in itertools.combinations(lines[p: p+25], 2):
        if a + b == lines[p + 25]:
            found = True
            break  # This break won't break me out of the outer loop, which is what I really want,
    if found:      # hence the 'found' check in the outer loop
        break      # writing it all in a function would also work - using `return` to jump out of the inner loop
```

Reply:
Aha. So `any` and `map` are lazy-evaluating when working on generators.

Reply:
Essentially, yeah 🦖